odoo/http.py

```diff
@@ -851,581 +1160,429 @@ def render(self, template, qcontext=None, lazy=True, **kw):
-        self.setdefault("session_token", None)
-        self.setdefault("context", {})
-        self.setdefault("debug", '')
-
-    def get_context(self):
+        Extract the params from the request's body and call the
+        endpoint. While it is prefered to override ir.http._pre_dispatch
+        and ir.http._post_dispatch, this method can be override to have
+        a tight control over the dispatching.
         """
-        Re-initializes the current user's session context (based on his
-        preferences) by calling res.users.get_context() with the old context.
 
-        :returns: the new context
+    def post_dispatch(self, response):
         """
-        assert self.uid, "The user needs to be logged-in to initialize his context"
-        self.context = dict(request.env['res.users'].context_get() or {})
-        self.context['uid'] = self.uid
-        self._fix_lang(self.context)
-        return self.context
+        Manipulate the HTTP response to inject various headers, also
+        save the session when it is dirty.
+        """
+        self.request._inject_future_response(response)
+        root.set_csp(response)
+
+        session = self.request.session
+        if session.can_save:
+            if session.should_save:
+                if session.rotate:
+                    root.session_store.delete(session)
+                    session.sid = root.session_store.generate_key()
+                    if session.uid:
+                        session.session_token = security.compute_session_token(session, self.request.env)
+                root.session_store.save(session)
+
+            # We must not set the cookie if the session id was specified
+            # using a http header or a GET parameter.
+            # There are two reasons to this:
+            # - When using one of those two means we consider that we are
+            #   overriding the cookie, which means creating a new session on
+            #   top of an already existing session and we don't want to
+            #   create a mess with the 'normal' session (the one using the
+            #   cookie). That is a special feature of the Javascript Session.
+            # - It could allow session fixation attacks.
+            if not self.request.session.explicit:
+                response.set_cookie('session_id', session.sid, max_age=SESSION_LIFETIME, httponly=True)
+
+    @abstractmethod
+    def handle_error(self, exc):
+        """
+        Transform the exception into a valid HTTP response. Called upon
+        any exception while serving a request.
+        """
+
 
-    def _fix_lang(self, context):
-        """ OpenERP provides languages which may not make sense and/or may not
-        be understood by the web client's libraries.
+class HttpDispatcher(Dispatcher):
+    routing_type = 'http'
 
-        Fix those here.
+    @classmethod
+    def is_compatible_with(cls, request):
+        return request.httprequest.content_type not in JSON_MIMETYPES
```

Comment:
```suggestion
        return request.httprequest.mimetype not in JSON_MIMETYPES
```
Mime type should be used instead of Content-Type as the later may also include other information like charset (eg. `Content-Type: application/json; charset=utf-8`), like in the mobile apps.

See for reference: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type#syntax